File: optimal_strategy_game.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends

long long dp[1001][1001];
long long dp2[1001][1001];
    
//Function to find the maximum possible amount of money we can win.
class Solution{
   
    int *a;
    long long solve(int i, int j) {
        if (i > j) return 0;
        if (dp[i][j] != -1) return dp[i][j];
        long long& ans = dp[i][j];
        ans = max(a[i] - solve2(i+1, j), a[j] - solve2(i, j-1));
        return ans;
    }
    
    long long solve2(int i, int j) {
        if (i == j) return a[i];
        if (dp[i][j] != -1) return dp[i][j];
        long long& ans = dp[i][j];
        ans = max(a[i] - solve(i+1, j), a[j] - solve(i, j-1));
        return ans;
    }
    
    
    public:
    long long maximumAmount(int arr[], int n){
        // Your code here
        memset(dp, -1, sizeof(dp));
        memset(dp2, -1, sizeof(dp2));
        a = arr;
        long long ans = solve(0, n-1);
        long long total = accumulate(arr, arr+n, 0LL);
        
        return (total - ans) / 2 + ans;
    }
};

//{ Driver Code Starts.
int main() 
{
   
   	int t;
	cin>>t;
	while(t--)
	{
		int n;
		cin>>n;
		int a[n];
		for(int i=0;i<n;i++)
		cin>>a[i];
		Solution ob;
		cout<< ob.maximumAmount(a,n)<<endl;
	}
	return 0;
}
// } Driver Code Ends

File: dp/optimal_strategy_game.py
# User function Template for python3
from functools import lru_cache
import sys

sys.setrecursionlimit(1000001)


# Function to find the maximum possible amount of money we can win.
class Solution:
    def optimalStrategyOfGame(self, n, arr):
        # code here
        @lru_cache(None)
        def f(i, j):
            if i == j:
                return arr[i]
            return max(arr[i] - f(i + 1, j), arr[j] - f(i, j - 1))

        return (sum(arr) + f(0, n - 1)) // 2


# {
# Driver Code Starts
# Initial Template for Python 3
import atexit
import io
import sys

# Contributed by : Nagendra Jha

if __name__ == "__main__":
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.optimalStrategyOfGame(n, arr))

# } Driver Code Ends
